sycl/include/sycl/detail/generic_type_traits.hpp

```diff
@@ -645,6 +633,18 @@ template <typename T, typename Enable = void> struct RelConverter {
   static R apply(value_t value) { return value; }
 };
 
+// TryToGetElementType<T>::type is T::element_type or T::value_type if those
+// exist, otherwise T.
+template <typename T> class TryToGetElementType {
+  static T check(...);
+  template <typename A> static typename A::element_type check(const A &);
+  template <typename A> static typename A::value_type check(const A &);
+
+public:
+  using type = decltype(check(T()));
+  static constexpr bool value = !std::is_same_v<T, type>;
+};
```

Comment:
Why do we keep `TryToGetElementType`?

Reply:
It has a single remaining use just below at line 650. Note that it's `::value`, not `::type` so I can't do the same replacement as in other places of this PR.

Reply:
I see. I personally think `value` in this is unintuitive, so I would prefer we replace it, if nothing else then by just replacing the use with `!std::is_same_v<T, get_elem_type_t<T>>`.

Reply:
It's not immediately obvious if that would be an NFC or not, so I won't be doing it here.